src/main/java/com/mixpanel/android/viewcrawler/ImageStore.java

```diff
@@ -0,0 +1,138 @@
+package com.mixpanel.android.viewcrawler;
+
+import android.content.Context;
+import android.content.res.Resources;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.graphics.drawable.BitmapDrawable;
+import android.graphics.drawable.Drawable;
+import android.util.Base64;
+import android.util.Log;
+
+import com.mixpanel.android.util.HttpService;
+import com.mixpanel.android.util.RemoteService;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * ABSOLUTELY NOT thread, or even process safe.
+ * Writes and reads files and directories at known paths, and uses a shared instance of MessageDigest.
+ */
+/* package */ class ImageStore {
+    public ImageStore(Context context) {
+        this(
+            context.getDir(DEFAULT_DIRECTORY_NAME, Context.MODE_PRIVATE),
+            new HttpService()
+        );
+    }
+
+    public ImageStore(File directory, RemoteService poster) {
+        mDirectory = directory;
+        mPoster = poster;
+        MessageDigest useDigest;
+        try {
+            useDigest = MessageDigest.getInstance("SHA1");
+        } catch (NoSuchAlgorithmException e) {
+            Log.w(LOGTAG, "Images won't be stored because this platform doesn't supply a SHA1 hash function");
+            useDigest = null;
+        }
+
+        mDigest = useDigest;
+    }
+
+    public Bitmap getImage(String url) {
+        final File file = storedFile(url);
+        byte[] bytes = null;
+        if (null == file || !file.exists()) {
+            try {
+                bytes = mPoster.performRequest(url, null);
+            } catch (IOException e) {
+                Log.i(LOGTAG, "Can't download bitmap", e);
+            } catch (RemoteService.ServiceUnavailableException e) {
+                Log.i(LOGTAG, "Couldn't download ");
+            }
+        }
+
+        final Bitmap bitmap;
+        if (null != bytes) {
+            if (null != file && bytes.length < MAX_BITMAP_SIZE) {
+                OutputStream out = null;
+                try {
+                    out = new FileOutputStream(file);
+                    out.write(bytes);
+                } catch (FileNotFoundException e) {
+                    Log.e(LOGTAG, "It appears that ImageStore is misconfigured, or disk storage is unavailable- can't write to bitmap directory", e);
+                } catch (IOException e) {
+                    Log.w(LOGTAG, "Can't store bitmap", e);
+                } finally {
+                    if (null != out) {
+                        try {
```

Comment:
It's still open- out.write could throw immediately, or even some random spot during the execution, and we want to make sure to close the stream. This is an annoying idiom, but otherwise we have to wait until GC happens to close the file if we open it and anything goes wrong.